crates/ra_hir_expand/src/db.rs

```diff
@@ -22,10 +34,16 @@ pub trait AstDatabase: SourceDatabase {
 
     #[salsa::interned]
     fn intern_macro(&self, macro_call: MacroCallLoc) -> MacroCallId;
-    fn macro_arg(&self, id: MacroCallId) -> Option<Arc<tt::Subtree>>;
-    fn macro_def(&self, id: MacroDefId) -> Option<Arc<mbe::MacroRules>>;
+    fn macro_arg(&self, id: MacroCallId) -> Option<(Arc<tt::Subtree>, Arc<mbe::TokenMap>)>;
+    fn macro_def(&self, id: MacroDefId) -> Option<(Arc<mbe::MacroRules>, Arc<mbe::TokenMap>)>;
     fn parse_macro(&self, macro_file: MacroFile) -> Option<Parse<SyntaxNode>>;
-    fn macro_expand(&self, macro_call: MacroCallId) -> Result<Arc<tt::Subtree>, String>;
+    fn parse_macro_with_info(&self, macro_file: MacroFile) -> Option<ParseMacroWithInfo>;
+    fn macro_expand(
+        &self,
+        macro_call: MacroCallId,
+    ) -> Result<(Arc<tt::Subtree>, MacroExpandInfo), String>;
```

Comment:
I think we can simplify the query structure here without loosing expressiveness. 

*First*, `macro_arg` and `macro_def` looks alright, no need to change those. *Second*, I think we can preserve an old `-> Result<Arc<tt::Subtree>>` signature of `macro_expand`. The caller can easily reconstruct both `TokenMap`s, as they have access to `MacroCallId`. In general, I think it sometimes makes sense to *project* a component of a pair in a query (for incremental), but just joining components of a pair doesn't seem better than making two original queries. If we restore the old `macro_expand` signature, we will be able to remove `MacroExpandInfo` type, which I think is good. *Third*, I think that rather introducing separate `parse_macro` and `parse_macro_with_info`, it would be sufficient to only have `fn parse_macro(&self, macro_file: MacroFile) -> Option<(Parse<SyntaxNode>, Arc<ExpansionInfo>)>;`. Or does this run into problems with incrementality, due to the fact that we store text rangs in `ExpansionInfo`?

Reply:
Thinking about the last bullet more, text ranges shouldn't be a problem, as the other component of the query value is a syntax tree, which is just as brittle, so a single `parse_macro` query should be fine.

Reply:
Yes, after refactor to remove the `MacroExpandInfo`, I just found that it is totally unnecessary. The expanded info is only used when some action is triggered. (e.g. goto-definition).